artifactory/services/utils/properties.go

```diff
@@ -9,92 +9,142 @@ import (
+	if _, exist := props.properties[key]; exist {
+		for _, existValue := range props.properties[key] {
+			if existValue == value {
+				return
+			}
+		}
 	}
-	// Remove leading semicolon
-	if strings.HasPrefix(encodedProps, ";") {
-		return encodedProps[1:]
+	props.properties[key] = append(props.properties[key], value)
+}
+
+func (props *Properties) ToEncodedString(concatValues bool) string {
+	encodedProps := ""
+	for key, values := range props.properties {
+		var jointProp string
+
+		if concatValues {
+			jointProp = fmt.Sprintf("%s=", url.QueryEscape(key))
+		}
+		for _, value := range values {
+			if concatValues {
+				propValue := strings.Replace(value, ValuesSeparator, fmt.Sprintf("\\%s", ValuesSeparator), -1)
+				jointProp = fmt.Sprintf("%s%s%s", jointProp, url.QueryEscape(propValue), url.QueryEscape(ValuesSeparator))
+			} else {
+				jointProp = fmt.Sprintf("%s%s=%s%s", jointProp, url.QueryEscape(key), url.QueryEscape(value), PropsSeparator)
+			}
+		}
+		// Trim the last comma/semicolon
+		if concatValues {
+			jointProp = strings.TrimSuffix(jointProp, url.QueryEscape(ValuesSeparator))
+		} else {
+			jointProp = strings.TrimSuffix(jointProp, PropsSeparator)
+		}
+
+		encodedProps = fmt.Sprintf("%s%s%s", encodedProps, PropsSeparator, jointProp)
 	}
-	return encodedProps
+	// Remove leading semicolon and return
+	return strings.TrimPrefix(encodedProps, PropsSeparator)
 }
 
 func (props *Properties) ToHeadersMap() map[string]string {
 	headers := map[string]string{}
-	for _, v := range props.Properties {
-		headers[v.Key] = base64.StdEncoding.EncodeToString([]byte(v.Value))
+	for key, values := range props.properties {
+		headers[key] = base64.StdEncoding.EncodeToString([]byte(strings.Join(values, ValuesSeparator)))
 	}
 	return headers
 }
 
 // Convert properties from Slice to map
 func (props *Properties) ToMap() map[string][]string {
-	propertiesMap := map[string][]string{}
-	for _, prop := range props.Properties {
-		propertiesMap[prop.Key] = append(propertiesMap[prop.Key], prop.Value)
+	return props.properties
+}
+
+func (props *Properties) KeysLen() int {
+	return len(props.properties)
+}
+
+func (props *Properties) removeDuplicateValues() {
+	for key, values := range props.properties {
+		props.properties[key] = removeDuplicates(values)
 	}
-	return propertiesMap
 }
 
-// Split properties string of format key=value to key value strings
-func splitProp(prop string) (string, string, error) {
-	splitIndex := strings.Index(prop, "=")
-	if splitIndex < 1 || len(prop[splitIndex+1:]) < 1 {
-		err := errorutils.CheckError(errors.New("Invalid property: " + prop))
-		return "", "", err
+func removeDuplicates(stringSlice []string) []string {
+	keys := make(map[string]bool)
+	var list []string
+
+	for _, item := range stringSlice {
+		if _, exist := keys[item]; !exist {
+			keys[item] = true
+			list = append(list, item)
+		}
+	}
+	return list
+}
+
+func MergeProperties(properties []*Properties) *Properties {
```

Comment:
Let's add a description for thus function.
Also, let's add a unit test for this function.